# **Task 1: Use NLP Techniques to Analyze a Collection of Texts**

## **Project Title: NLP-Based Topic Modeling of Consumer Complaints**

### **Notebook 2: Text Vectorization**

#### **Introduction**

After cleaning and preprocessing the dataset in the previous step, the next essential step is text vectorization. Since machine learning models cannot directly process raw text, we must convert it into a structured numerical format. This notebook focuses on transforming the preprocessed text data into numerical representations using two popular vectorization techniques: Bag of Words (BoW) and Term Frequency-Inverse Document Frequency (TF-IDF).

These vectorized representations will serve as input for the topic modeling algorithms applied in later notebooks.

## **1. Import Necessary Libraries**

In [1]:
import pandas as pd
import numpy as np

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

import warnings
warnings.filterwarnings("ignore")

## **2. Load Preprocessed Data**

In [2]:
# Define the file path
file_path = "/Users/achshahrm/Library/Mobile Documents/com~apple~CloudDocs/Documents/IU International/Sem 05/Project Data Analysis/project_final/data/preprocessed_complaints.csv"

# Load the data into a DataFrame
data_preprocessed = pd.read_csv(file_path)

# Display the first few rows of the DataFrame
data_preprocessed.head()

,complaints_text
0,purchase order day shipping amount receive pro...
1,forward message date tue subject please invest...
2,forward message cc send friday pdt subject fin...
3,payment history miss credit report specialize ...
4,payment history miss credit report make mistak...


## **3. BoW - Bag of Words**

In [3]:
# Initialize CountVectorizer for Bag of Words
bow_vectorizer = CountVectorizer(max_features=5000, stop_words='english')

# Fit and transform the complaints_text column
bow_matrix = bow_vectorizer.fit_transform(data_preprocessed['complaints_text'])

# Convert to DataFrame for easier inspection
bow_df = pd.DataFrame(bow_matrix.toarray(), columns=bow_vectorizer.get_feature_names_out())
bow_df.head()

,aa,aadvantage,aag,aargon,abandon,abide,ability,able,abroad,abruptly,...,yo,york,young,yr,zelle,zero,zip,zombie,zone,zwicker
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,1,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,2,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,2,0,0,...,0,0,0,0,0,0,0,0,0,0


## **4. TF-IDF - Term Frequency Inverse Document Frequency**

In [4]:
# Initialize TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(max_features=5000, stop_words='english')

# Fit and transform the complaints_text column
tfidf_matrix = tfidf_vectorizer.fit_transform(data_preprocessed['complaints_text'])

# Convert to DataFrame for easier inspection
tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=tfidf_vectorizer.get_feature_names_out())
tfidf_df.head()

,aa,aadvantage,aag,aargon,abandon,abide,ability,able,abroad,abruptly,...,yo,york,young,yr,zelle,zero,zip,zombie,zone,zwicker
0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.0,0.0,...,0.0,0.0000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.0,0.0,...,0.0,0.0968,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.0,0.0,...,0.0,0.0000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.105765,0.0,0.0,...,0.0,0.0000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.117352,0.0,0.0,...,0.0,0.0000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


## **5. Save Results**

In [10]:
# Save BoW and TF-IDF matrices
bow_df.to_csv("/Users/achshahrm/Library/Mobile Documents/com~apple~CloudDocs/Documents/IU International/Sem 05/Project Data Analysis/project_final/results/vectorized/bow_representation.csv", index=False)
tfidf_df.to_csv("/Users/achshahrm/Library/Mobile Documents/com~apple~CloudDocs/Documents/IU International/Sem 05/Project Data Analysis/project/results/vectorized/tfidf_representation.csv", index=False)

# Save feature names
np.savetxt("/Users/achshahrm/Library/Mobile Documents/com~apple~CloudDocs/Documents/IU International/Sem 05/Project Data Analysis/project_final/results/vectorized/feature_names_bow.txt", bow_vectorizer.get_feature_names_out(), fmt="%s")
np.savetxt("/Users/achshahrm/Library/Mobile Documents/com~apple~CloudDocs/Documents/IU International/Sem 05/Project Data Analysis/project_final/results/vectorized/feature_names_tfidf.txt", tfidf_vectorizer.get_feature_names_out(), fmt="%s")

print("BoW and TF-IDF representations along with feature names saved.")

BoW and TF-IDF representations along with feature names saved.


## **6. Compare Results**

### Compare Shape of Matrices

In [6]:
print("BoW Shape:", bow_df.shape)
print("TF-IDF Shape:", tfidf_df.shape)

BoW Shape: (10000, 5000)
TF-IDF Shape: (10000, 5000)


Both methods produce matrices of the same shape since the same dataset and vocabulary size were used. The primary difference lies in how the values in the matrix are calculated. In BoW, the matrix values represent raw word counts, whereas, in TF-IDF, the matrix values are normalized term weights, emphasizing unique and important words while down-weighting common words.

### Inspect Word Frequency Distributions

In [7]:
# Sum word frequencies for BoW and TF-IDF
bow_word_freq = bow_df.sum(axis=0).sort_values(ascending=False)
tfidf_word_freq = tfidf_df.sum(axis=0).sort_values(ascending=False)

print("Top 10 Words in BoW:")
print(bow_word_freq.head(10))

print("Top 10 Words in TF-IDF:")
print(tfidf_word_freq.head(10))

Top 10 Words in BoW:
account        21302
credit         17757
report         15885
payment        10326
information     6715
pay             6591
card            6251
bank            6133
time            6074
debt            5913
dtype: int64
Top 10 Words in TF-IDF:
account        745.975098
credit         670.444864
report         665.845518
payment        410.720553
debt           337.184940
card           320.170812
bank           311.677124
company        309.167602
pay            302.822093
information    283.453997
dtype: float64


The lists above highlight the top 10 words in the dataset based on BoW (Bag of Words) and TF-IDF (Term Frequency-Inverse Document Frequency) representations:

1. **Top Words in BoW**:
   - Words like account, credit, and report have the highest raw counts, indicating they are the most frequently used words in the dataset.
   - BoW emphasizes the frequency of words without considering their importance or uniqueness across documents.

2. **Top Words in TF-IDF**:
   - The same words appear in TF-IDF with different weights, where the values represent the relative importance of each word in the corpus.
   - Words like company and debt are emphasized more in TF-IDF, as they are likely important within certain documents but not overly common across the entire dataset.
   - TF-IDF down-weights words that are frequent across all documents (e.g., account still ranks high, but its weight is normalized compared to its BoW count).

### Compare Sparsity

In [8]:
# Calculate sparsity for BoW and TF-IDF
bow_sparsity = 1.0 - (bow_matrix.nnz / (bow_matrix.shape[0] * bow_matrix.shape[1]))
tfidf_sparsity = 1.0 - (tfidf_matrix.nnz / (tfidf_matrix.shape[0] * tfidf_matrix.shape[1]))

print(f"BoW Sparsity: {bow_sparsity:.4f}")
print(f"TF-IDF Sparsity: {tfidf_sparsity:.4f}")

BoW Sparsity: 0.9911
TF-IDF Sparsity: 0.9911


The sparsity of the Bag of Words (BoW) and TF-IDF matrices is 0.9911, meaning that approximately 99.11% of the matrix values are zero. This high sparsity is common in text data.

## **Next Steps**
Now that the text has been numerically transformed, we proceed to the next stage, Finding the optimal number of topics (k) by evaluating topic coherence scores across different models.